compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -2294,6 +2454,29 @@ object Types {
     myHash = hc
   }
 
+  /** A `TermRef` that, through flow-sensitive type inference, we know is non-null.
+   *  Accordingly, the `info` in its denotation won't be of the form `T|Null`.
+   *  Notice that this class isn't cached, unlike the regular `TermRef`.
+   */
+  final class NonNullTermRef(prefix: Type, designator: Designator) extends TermRef(prefix, designator) {
+    // There's nothing special about this class: it's just used as a marker to identify certain
+    // `TermRef`s in `computeDenot`.
+  }
```

Comment:
In fact, I don't think this will work. Since `TermRef` is declared as cached, so it has to satisfy that protocol. But `NonNullTermRef` is not cached. This will cause problems. Probably only in rare situations, but when they arise they will be impossibly hard to debug.

Reply:
It might be better to work on the denotation instead. Use a normal `TermRef`, but give it a special denotation that strips nulls.